[1단계 - 행운의 로또 게임] 다이앤(고다은) 미션 제출합니다

Daeun-100 opened the pull request:
안녕하세요 온스타! 이번 페어프로그래밍은 3인(with 데이지🌸 세라🍀)와 함께 진행하였습니다. 리뷰 잘 부탁드립니다.😊
## 학습 목표
이번 미션을 통해 다음과 같은 학습 경험들을 쌓는 것을 목표로 합니다.
- UI와 도메인 영역을 분리할 수 있는 설계를 고민해보고, 목적에 맞게 객체와 함수를 활용
- TDD 방식으로 개발하며, 단위 테스트 기반으로 점진적인 리팩터링

## 제출 전 체크 리스트  
<!-- 리뷰를 요청하기 전, 다음 항목들을 기본적으로 확인해 주세요. -->
- [x] 기능 요구 사항을 모두 구현했고, 정상적으로 동작하는지 확인했나요?
- [x] 기본적인 프로그래밍 요구 사항을 준수하고 있는지 확인했나요?
- [x] 테스트 코드는 모두 정상적으로 실행되나요?
- [x] (해당하는 경우) 배포한 데모 페이지에 정상적으로 접근할 수 있나요?
  - 배포한 링크 기입: 

## 리뷰 요청 & 논의하고 싶은 내용  
<!-- PR 작성자로서, 코드만으로는 알기 어려운 작성자의 의도와 문제 해결 과정에 대해 공유해 주세요.  
이 PR에서 달성해야 하는 학습 목표를 잘 달성하고 있는지 스스로 확인하며, 피드백 받을 수 있는 내용일지 점검해 보세요.
리뷰어에게 정답을 묻기보다 고민하고 의사 결정한 과정에 대해 공유하고 이에 대한 피드백을 받으며 대화해보기를 권장합니다. -->


### 1) 이번 단계에서 가장 많이 고민했던 문제와 해결 과정에서 배운 점
<!-- 구현 과정에서 가장 어려웠던 점이나 많이 고민한 점은 무엇인가요?
이를 해결하기 위해 어떤 방법들을 검토하고 시도했으며, 그 과정에서 새롭게 배운 점이 있나요? --> 
- 이번 미션에서는 각 도메인을 class/객체/함수 중 어떤 형식으로 만들지 고민을 했고 아래와 같은 기준을 세웠습니다. 
  `Lotto` :  `class` , instance를 여러개 만들 수도 있어서 객체로 만듬
  `LottoResult` : `class` , 전체 당첨 내역과 수익률을 구하기 위해서는 공통된 winningHistory 필드가 필요 
  `LottoMachine` : `class` , 발행된 로또를 모아두는 곳이 필요함
  `Ticket` :  `객체` ticket 발행하는 역할을 담당, 필드 필요x
 `input,,output` : `객체` , 입력,출력 하는 역할을 담당, 필드 필요x
이렇게 판단을 하고 각 도메인을 위 형식으로 만들었는데 이러한 기준에 대해 어떻게 생각하시나요? 온스타는 class/객체/함수를 나누는 기준이 있나요?

- 저희는 역할에 따라 domain을 분리해 봤습니다. 최대한 비슷한 역할을 하는 코드가 같이 있도록 짜봤는데 잘 구분이 되었는지 궁금합니다
 `Lotto` : 숫자확인 로직들
 `LottoResult` : 당첨내역/수익률 계산
 `LottoMachine` : 발행된 로또들 저장, 발행한 로또들을 돌면서 숫자 확인하고 순위 반환

### 2) 이번 리뷰를 통해 논의하고 싶은 부분  
<!-- 구현한 코드와 학습 목표와 관련해 피드백을 받고 싶은 부분이나, 함께 논의해보고 싶은 점 -->
- 페어 프로그래밍을 하면서 메소드 선언 순서에 대해 의견이 갈습니다. 저는 코드의 가독성을 높이기 위해, 메소드가 호출되는 흐름에 따라 선언하는 것이 좋다고 생각했습니다. 예를 들어, a 함수와 b 함수가 존재하고, 이 둘을 사용하는 c 함수가 있다면, c 함수를 후순위에 두는 것이 더 자연스럽다고 판단했습니다 다른 페어는 핵심 로직을 위에 작성하는 것이 좋다고 판단했고 논의한 끝에 호출되는 흐름에 따라 쓰기로 결정했습니다. 온스타는 어떻게 생각하시나요?

- TDD를 통해 작은 단위로 문제를 바라볼 수 잇다는 것은 좋은 점이라고 느꼈습니다. 하지만 도메인에 대해 숙지 되지 않은 상태에서 TDD를 사용하다 보니 시간이 더 오래 걸리고 시간대비 어느 장점이 있는지 와닿지 않는 것 같습니다. 기능 개발 후 테스트를 작성하는 것과의 뚜렷한 장점도 잘 느끼지 못한 것 같습니다.

- 첫번째 미션에서도 질문했던 거긴 하지만 아직 단위 테스트의 기준이 뭔지 잘 모르겠습니다. 단위테스트는 독립적으로 테스트가 가능해야 한다고 생각하는데, 지금은 한 객체를 테스트하는데 다른 객체가 필요해서 테스트가 독립적으로 작동하지 않는다고 생각해요. 온스타는 어떻게 생각하시나요?
---

<details>
<summary>✅ 리뷰어 체크 포인트</summary>
<!-- 리뷰어가 이 PR을 검토할 때 중점적으로 확인할 사항입니다.-->

### 1단계
- [ ] TDD를 활용해 기능을 구현하는 과정에서 적절한 테스트 우선 접근 방식을 적용했는가? 단위 테스트의 커버리지는 충분한가?
- [ ] 도메인과 UI의 관심사를 분리하여 적절한 모듈화가 이루어졌는가? 하나의 객체나 모듈이 너무 많은 책임을 가지고 있지는 않은가?
- [ ] 객체의 프로퍼티를 직접 조작하기보다 메시지를 던지고 있는가?
- [ ] 불필요한 클래스를 사용하지 않고, 함수를 적극적으로 활용하여 JavaScript다운 방식으로 로직을 구현했는가?

### 2단계 
- [ ] 도메인 로직에 불필요한 영향을 주지 않고 UI 변경에 대응했는가?
- [ ] DOM 조작과 이벤트 활용을 JavaScript의 개념에 맞게 이해하고 적절하게 적용했는가? 
- [ ] 웹 표준을 준수하는 마크업을 활용하며, 스타일 작성에 일관성이 있는가?
</details>

keemsebin committed:
chore: 기능목록 readme 작성

Co-Authored-By: Daeun-100 <Daeun-100@users.noreply.github.com>
Co-Authored-By: Yugyeong Kim <kimyou1102@users.noreply.github.com>

keemsebin committed:
feat: 입력받은 금액 만큼 로또 개수 생성

Co-authored-by: Daeun-100 <Daeun-100@users.noreply.github.com>
Co-authored-by: Yugyeong Kim <kimyou1102@users.noreply.github.com>

keemsebin committed:
feat: 정해진 개수만큼 로또 생성 함수 구현

Co-authored-by: Daeun-100 <Daeun-100@users.noreply.github.com>
Co-authored-by: Yugyeong Kim <kimyou1102@users.noreply.github.com>

keemsebin committed:
feat: random 값 생성 함수 구현, 중복되지 않은 랜덤 값을 담은 배열 함수 구현

Co-authored-by: Daeun-100 <Daeun-100@users.noreply.github.com>
Co-authored-by: Yugyeong Kim <kimyou1102@users.noreply.github.com>

keemsebin committed:
feat: Lotto 번호 주입, 로또 번호 get 함수 구현

Co-authored-by: Daeun-100 <Daeun-100@users.noreply.github.com>
Co-authored-by: Yugyeong Kim <kimyou1102@users.noreply.github.com>

keemsebin committed:
feat: 로또 생성 시 랜덤 번호 주입 추가

Co-authored-by: Daeun-100 <Daeun-100@users.noreply.github.com>
Co-authored-by: Yugyeong Kim <kimyou1102@users.noreply.github.com>

keemsebin committed:
feat: 로또 번호 오름차순 정렬 메소드 구현

Co-authored-by: Daeun-100 <Daeun-100@users.noreply.github.com>
Co-authored-by: Yugyeong Kim <kimyou1102@users.noreply.github.com>

keemsebin committed:
test: 로또 번호 오름차순 테스트 추가

Co-authored-by: Daeun-100 <Daeun-100@users.noreply.github.com>
Co-authored-by: Yugyeong Kim <kimyou1102@users.noreply.github.com>

keemsebin committed:
feat: 기본 입력 함수 구현 구현

Co-authored-by: Daeun-100 <Daeun-100@users.noreply.github.com>
Co-authored-by: Yugyeong Kim <kimyou1102@users.noreply.github.com>

keemsebin committed:
feat: 랜덤으로 주어진 로또 번호와 입력받은 로또 번호의 일치 갯수 함수 구현

Co-authored-by: Yugyeong Kim <kimyou1102@users.noreply.github.com>
Co-authored-by: Daeun-100 <Daeun-100@users.noreply.github.com>

keemsebin committed:
feat: 로또에 보너스 번호 존재 여부 확인 함수 구현

Co-authored-by: Yugyeong Kim <kimyou1102@users.noreply.github.com>
Co-authored-by: Daeun-100 <Daeun-100@users.noreply.github.com>

keemsebin committed:
feat: 발행한 로또 번호와 입력한 로또 번호/보너스 번호의 일치 갯수를 기반으로 로또 상태를 매칭하는 함수 구현

Co-authored-by: Daeun-100 <Daeun-100@users.noreply.github.com>
Co-authored-by: Yugyeong Kim <kimyou1102@users.noreply.github.com>

keemsebin committed:
feat: 당첨내역을 기반으로 총 수입을 구하는 메소드 구현

Co-authored-by: Yugyeong Kim <kimyou1102@users.noreply.github.com>
Co-authored-by: Daeun-100 <Daeun-100@users.noreply.github.com>

keemsebin committed:
feat: 당첨 내역을 기반으로 수익률을 계산하는 메소드 구현

Co-authored-by: Yugyeong Kim <kimyou1102@users.noreply.github.com>
Co-authored-by: Daeun-100 <Daeun-100@users.noreply.github.com>

keemsebin committed:
fix: createLotto 빈배열 반환하는 문제 해결

Co-authored-by: Yugyeong Kim <kimyou1102@users.noreply.github.com>
Co-authored-by: Daeun-100 <Daeun-100@users.noreply.github.com>

keemsebin committed:
fix: getMatchedLottoRank 함수 매개변수에 bonusNumber 추가

Co-authored-by: Daeun-100 <Daeun-100@users.noreply.github.com>
Co-authored-by: Yugyeong Kim <kimyou1102@users.noreply.github.com>

keemsebin committed:
feat: 로또 구입 금액 입력 로직 추가

Co-authored-by: Yugyeong Kim <kimyou1102@users.noreply.github.com>
Co-authored-by: Daeun-100 <Daeun-100@users.noreply.github.com>

keemsebin committed:
feat: INPUT, OUTPUT 상수화

Co-authored-by: Daeun-100 <Daeun-100@users.noreply.github.com>
Co-authored-by: Yugyeong Kim <kimyou1102@users.noreply.github.com>

keemsebin committed:
feat: index.js에 당첨번호, 보너스 번호, 재입력 로직 추가

Co-authored-by: Daeun-100 <Daeun-100@users.noreply.github.com>
Co-authored-by: Yugyeong Kim <kimyou1102@users.noreply.github.com>

keemsebin committed:
chore: readme 유효성 검사 케이스 추가

Co-authored-by: Daeun-100 <Daeun-100@users.noreply.github.com>
Co-authored-by: Yugyeong Kim <kimyou1102@users.noreply.github.com>

keemsebin committed:
feat: 출력 기능을 담당하는 객체 생성

Co-authored-by: Daeun-100 <Daeun-100@users.noreply.github.com>
Co-authored-by: Yugyeong Kim <kimyou1102@users.noreply.github.com>

keemsebin committed:
feat: Validator class 구현

Co-authored-by: Yugyeong Kim <kimyou1102@users.noreply.github.com>
Co-authored-by: Daeun-100 <Daeun-100@users.noreply.github.com>

keemsebin committed:
feat: 에러메세지 constants 정의

Co-authored-by: Yugyeong Kim <kimyou1102@users.noreply.github.com>
Co-authored-by: Daeun-100 <Daeun-100@users.noreply.github.com>

keemsebin committed:
feat: 에러 throw utils 함수 구현

Co-authored-by: Yugyeong Kim <kimyou1102@users.noreply.github.com>
Co-authored-by: Daeun-100 <Daeun-100@users.noreply.github.com>

keemsebin committed:
feat: Lotto class 내 validate 메소드 추가

Co-authored-by: Yugyeong Kim <kimyou1102@users.noreply.github.com>
Co-authored-by: Daeun-100 <Daeun-100@users.noreply.github.com>

keemsebin committed:
feat: lotto validator class 생성, lotto validator test 작성

Co-authored-by: Daeun-100 <Daeun-100@users.noreply.github.com>
Co-authored-by: Yugyeong Kim <kimyou1102@users.noreply.github.com>

keemsebin committed:
feat: 보너스 번호 validator class 구현, test 작성

Co-authored-by: Daeun-100 <Daeun-100@users.noreply.github.com>
Co-authored-by: Yugyeong Kim <kimyou1102@users.noreply.github.com>

keemsebin committed:
feat: 구매한 금액에 대한 유효성 검사 class 생성, test 작성

Co-authored-by: Daeun-100 <Daeun-100@users.noreply.github.com>
Co-authored-by: Yugyeong Kim <kimyou1102@users.noreply.github.com>

keemsebin committed:
feat: 로또 번호가 6개 미만인 경우 validation 조건 추가, test 작성

Co-authored-by: Daeun-100 <Daeun-100@users.noreply.github.com>
Co-authored-by: Yugyeong Kim <kimyou1102@users.noreply.github.com>

keemsebin committed:
feat: 전체 로직 index.js에 추가

Co-authored-by: Yugyeong Kim <kimyou1102@users.noreply.github.com>
Co-authored-by: Daeun-100 <Daeun-100@users.noreply.github.com>

keemsebin committed:
fix: printLottoNumber 출력 시 lotto number 받아오는 함수 추가

keemsebin committed:
fix: retry 함수에 Input retry 객체를 다시 받아오도록 수정

keemsebin committed:
refactor: createLotto 함수를 Ticket 객체 매소드로 변경

Co-authored-by: Yugyeong Kim <kimyou1102@users.noreply.github.com>
Co-authored-by: Daeun-100 <Daeun-100@users.noreply.github.com>

keemsebin committed:
refactor: count 함수명 변경

keemsebin committed:
chore: 사용하지 않는 import 제거

keemsebin committed:
feat: 상수 Object freeze 추가

Co-authored-by: Yugyeong Kim <kimyou1102@users.noreply.github.com>
Co-authored-by: Daeun-100 <Daeun-100@users.noreply.github.com>

keemsebin committed:
refactor: 일치하는 로또 상태를 반환하는 메소드명 변경

keemsebin committed:
refactor: updateFinalStatus 메소드 추가, getMatchedLottoStatus 메소드 간소화

Co-authored-by: Yugyeong Kim <kimyou1102@users.noreply.github.com>
Co-authored-by: Daeun-100 <Daeun-100@users.noreply.github.com>

keemsebin committed:
feat: 당첨 통계, 라인 출력 문자 추가

Co-authored-by: Yugyeong Kim <kimyou1102@users.noreply.github.com>
Co-authored-by: Daeun-100 <Daeun-100@users.noreply.github.com>

keemsebin committed:
refactor: Ticket 객체에 로또 발행 함수 추가

Co-authored-by: Yugyeong Kim <kimyou1102@users.noreply.github.com>
Co-authored-by: Daeun-100 <Daeun-100@users.noreply.github.com>

keemsebin committed:
refactor: index.js Input 로직 분리

Co-authored-by: Yugyeong Kim <kimyou1102@users.noreply.github.com>
Co-authored-by: Daeun-100 <Daeun-100@users.noreply.github.com>

keemsebin committed:
chore: readme 폴더 구조 추가

keemsebin committed:
fix: printInput 함수 화살표 함수 누락 추가

Co-authored-by: Yugyeong Kim <kimyou1102@users.noreply.github.com>
Co-authored-by: Daeun-100 <Daeun-100@users.noreply.github.com>

onschan commented on __tests__/BonusNumberValidator.test.js:
```diff
@@ -0,0 +1,13 @@
+
+    expect(() =>
+      bonusValidator.validateBonusNumber(lottoNumbers, bonusNumber)
+    ).toThrow("[ERROR]");
```
하드코딩된 "ERROR"를 테스트 코드에 사용하는 부분이 테스트의 신뢰도를 떨어뜨리고있는 것 같아요.
상수화된 값을 사용해주시고, 에러 메시지를 상수로 관리하고 있으신 만큼 정확한 에러 메시지가 테스트 코드에 포함된다면 신뢰도가 더 올라갈 것 같아요.

onschan commented on __tests__/Lotto.test.js:
```diff
@@ -0,0 +1,74 @@
+
+  // then
+  expect(lottos.length).toBe(lottoCount);
+});
```
해당 테스트들은 요구사항의 시나리오상 로또와 연관성이 있을뿐 Lotto 객체를 테스트하는 부분에는 어울리지 않는 것 같아요

onschan commented on __tests__/Lotto.test.js:
```diff
@@ -0,0 +1,74 @@
+  const bonusNumber = 3;
+
+  expect(lotto.hasBonusNumber(bonusNumber)).toBeTruthy();
+});
```
describe를 활용해서 해당 클래스의 메서드들을 놓치지않고 테스트하고있는지 그룹화를 해보는 것도 좋을 것 같아요.

onschan commented on __tests__/LottoMachine.test.js:
```diff
@@ -0,0 +1,34 @@
+    // then
+    expect(
+      machine.getMatchedLottoStatus(enteredLottoNumbers, bonusNumber)
+    ).toEqual([{ RANK: 2, COUNT: 5, REWORD: 30_000_000, IS_BONUS: true }]);
```
COUNT, REWORD, IS_BONUS 등은 2등을 반환한다는 시나리오와 크게 관련은 없지만 테스트 코드와 강하게 연결되어있네요.
필요한 데이터가 추가되거나 세부 요구 사항이 변경될때마다 해당 테스트는 깨질 가능성이 클 것 같아요.
machine.getMatchedLottoStatus의 RANK에 대해서만 테스트해볼 수 있지않을까요?

onschan commented on src/domain/Lotto.js:
```diff
@@ -0,0 +1,32 @@
+
+  hasBonusNumber(bonusNumber) {
+    return this.numbers.includes(bonusNumber);
+  }
```
getSameNumbers, hasBonusNumber 등 당첨 결과를 위해 존재하는 메서드들은 단일 로또 객체의 역할로 볼 수 있을까요? 🤔
제 생각엔 로또 번호와 당첨 번호, 보너스 번호를 비교하는 LottoMachine의 역할과 가까워보이는데, 로또가 해당 메서드를 가지고있다보니 LottoMachine의 구현과 테스트에서 강하게 결합되는 구조를 가졌던게 아닐까해요.

onschan commented on __tests__/LottoMachine.test.js:
```diff
@@ -0,0 +1,34 @@
+describe("발행한 로또 번호와 입력한 로또 번호의 일치 갯수를 기반으로 로또 상태를 매칭한다", () => {
+  test("숫자 5개가 일치하고 보너스 숫자가 존재하는 경우 2등을 반환한다.", () => {
+    // given
+    const issuedLottoNumbers = [new Lotto([2, 3, 4, 5, 6, 7])];
```
테스트 코드부터 Lotto의 구현체가 바뀌면 LottoMachine의 테스트에도 영향을 줄 수 밖에 없게 되어있어요.
LottoMachine과 Lotto가 너무 강하게 의존되어있다는 부분부터 불편함을 느끼시면 좋아요.

LottoMachine 코멘트에 이어서 추가할게요.

onschan commented on src/domain/LottoResult.js:
```diff
@@ -0,0 +1,23 @@
+  }
+
+  getRate() {
+    return ((this.getTotalProfit() / this.price) * 100).toFixed(1);
```
상황에 따라 다를 수 있지만 toFixed 메서드의 특징으로 문자열로 추출되는 부분도 number 형으로 변환해서 반환하는게 좋다고 생각해요.
도메인에서 반환하는 값은 추가 연산 및 비교에도 이용될 수 있어요.

onschan commented on __tests__/LottoResult.test.js:
해당 테스트도 describe가 있으면 좋을 것같아요.

onschan commented on __tests__/LottoResult.test.js:
```diff
@@ -0,0 +1,46 @@
+
+  //then
+  expect(profit).toEqual(30_050_000);
+});
```
getTotalProfit 메서드에 대한 테스트로 당첨 내역이 없는 경우도 테스트가 진행되는 편이 좋을 것 같아요.

onschan commented on src/constants/message.js:
```diff
@@ -0,0 +1,27 @@
+  INCLUDE: "보너스 번호는 로또 번호와 중복될 수 없습니다.",
+  UNIT: "구입 금액은 1000원 단위로 입력해야 합니다.",
+  INVALID_RETRY_STRING: "y 또는 n을 입력해주세요.",
+  LENGTH: "로또 번호는 6개여야 합니다.",
```
상수 정리를 잘 해주시고 계시는 만큼 출력 문구의 매직넘버도 상수로 관리하는게 어떨까요?

onschan commented on __tests__/LottoValidator.test.js:
```diff
@@ -0,0 +1,24 @@
+    const lottoValidator = new LottoValidator();
+
+    expect(() => lottoValidator.validateLotto(lottoNumber)).toThrow("[ERROR]");
+  });
```
LOTTO_NUMBER.LENGTH 를 활용해서 해당 값으로 도메인, 테스트를 컨트롤할 수 있지않을까요?

Daeun-100 replied:
해당 값을 활용해서 도메인,테스트를 컨트롤 한다는게 정확히 어떤 의미인지 이해하기 어렵습니다!
LOTTO_NUMBER.LENGTH를 사용해 숫자 배열을 생성하고 test 설명의 매직넘버를 변경했는데 제가 이해한게 맞을까요? 
https://github.com/woowacourse/javascript-lotto/pull/360/commits/d9730786ece01395de836e7801b34c9fc858da94

onschan replied:
네 맞습니다. 상수로 관리되는 값이 변경되어도 테스트 코드가 깨지지않도록 요청드렸던 부분이었고 잘 반영해주셨어요

onschan commented on __tests__/PriceValidator.test.js:
```diff
@@ -0,0 +1,20 @@
+
+      expect(() => priceValidator.validatePrice(price)).toThrow("[ERROR]");
+    }
+  );
```
999, 100001 처럼 경계값을 기준으로 잡아서 테스트해볼까요?

onschan commented on __tests__/PriceValidator.test.js:
```diff
@@ -0,0 +1,20 @@
+  });
+
+  test.each([500, 1000000])(
+    "1000원 이상 100000 이하가 아닌 경우 에러를 발생시킨다. - %s원인 경우",
```
```suggestion
    "1000원 이상 100,000 이하가 아닌 경우 에러를 발생시킨다. - %s원인 경우",
```
사소한 습관이지만 도움이 많이 될거에요!

onschan commented on __tests__/Validator.test.js:
```diff
@@ -0,0 +1,32 @@
+    expect(() => validator.range({ min: 1, max: 45 }, input)).toThrow(
+      "[ERROR]"
+    );
+  });
```
이 부분도 1, 45의 두 경계값을 기준으로 테스트해주세요

onschan commented on readme.md:
```diff
@@ -0,0 +1,115 @@
+    - [x]  예외3) 범위를 벗어난 경우
+    - [x]  예외4) 로또번호와 중복되는 경우
+- [ ] 재시도 여부
+    - [x] 예외1) y또는 n이 아닌 문자를 입력한 경우
```
잘 정리해주셨는데요. 예외를 기준으로 작성하다보면 놓치는 케이스가 발생할 수 있어요. 정상 플로우를 중심으로 작성하고 구현 단계에서 예외에 대해서 철저하게 고민하고 잡아주시는게 편하실거에요.

onschan commented on src/constants/lotto.js:
```diff
@@ -0,0 +1,13 @@
+  { RANK: 2, COUNT: 5, REWORD: 30_000_000, IS_BONUS: true },
+  { RANK: 3, COUNT: 5, REWORD: 1_500_000, IS_BONUS: false },
+  { RANK: 4, COUNT: 4, REWORD: 50_000, IS_BONUS: false },
+  { RANK: 5, COUNT: 3, REWORD: 5000, IS_BONUS: false },
```
큰 숫자에 대한 가독성도 잘 챙겨주셨네요 👍

onschan commented on src/domain/Lotto.js:
```diff
@@ -0,0 +1,32 @@
+
+class Lotto {
+  constructor(numbers) {
+    this.numbers = this.sortLottoNumber(numbers);
```
인스턴스의 numbers 필드에 직접 접근할 수 없도록 private 필드로 만들어볼까요?

onschan commented on src/domain/LottoMachine.js:
```diff
@@ -0,0 +1,53 @@
+    return this.issuedLottoNumbers.map((lotto) => {
+      return lotto.hasBonusNumber(bonusLottoNumbers);
+    });
+  }
```
LottoMachine의 메서드가 Lotto의 구현체(getSameNumbers, hasBonusNumber)에 의존하고 있는 부분있는데, 배드 패턴이라고 규정할 수는 없으나 결합도는 낮출 수 있지않을까 싶어요.
issuedLottoNumbers가 Lotto의 인스턴스가 아닌 [[1,2,3,4,5,6], [2,3,4,5,6,7]] 의 "값"만 가지고있다면 해당 클래스의 역할을 하는데 충분하지 않을까요?

onschan commented on src/domain/LottoMachine.js:
```diff
@@ -0,0 +1,53 @@
+import { LOTTO_STATUS } from "../constants/lotto.js";
+
+class LottoMachine {
+  issuedLottoNumbers;
```
```suggestion
  #issuedLottoNumbers;
  #matchedLottoStatus;
```

onschan commented on src/domain/LottoMachine.js:
```diff
@@ -0,0 +1,53 @@
+  updateStatus(callback) {
+    const currentStatus = LOTTO_STATUS.find(callback);
+    this.matchedLottoStatus.push(currentStatus);
+  }
```
updateStatus의 callback 인자와 동작이 메서드명만 본다면 구현체를 예측하기 힘들 것 같아요 🤔

`const currentStatus = LOTTO_STATUS.find(callback);` 이 부분은 해당 메서드 외부에 일임시키고
"Status를 추가해준다" 라는 가벼운 역할만 수행하고 예측 가능하도록 메서드 명과 동작을 수정해볼까요?

onschan commented on src/domain/LottoMachine.js:
```diff
@@ -0,0 +1,53 @@
+    this.updateFinalStatus(matchingNumbers, isBonusArray);
+
+    return this.matchedLottoStatus;
+  }
```
getMatchedLottoStatus 는 this.matchedLottoStatus만 반환하는 getter 역할만 수행하도록 하고 그 외의 동작들은 외부로 분리해주시는게 좋을 것 같아요.

onschan commented on src/domain/Ticket.js:
```diff
@@ -0,0 +1,28 @@
+  checkAndAddLottoNumbers: (store, number) => {
+    if (store.has(number)) return;
+    store.add(number);
+  },
```
Set 자료 구조형은 중복을 허용하지 않기때문에 해당 메서드 없이 store.add 만해주어도 충분하겟네요!

onschan commented on src/step1-index.js:
```diff
@@ -1,4 +1,85 @@
+  const retry = await Input.retry(async () => {
+    const input = await Input.readLineAsync(INPUT.RETRY);
+    if (!RETRY_STRING.includes(input)) {
+      throwError(ERROR);
```
이 부분은 놓치신걸까요?

onschan commented on src/step1-index.js:
```diff
@@ -1,4 +1,85 @@
+    return input.toLowerCase();
+  });
+
+  if (retry === "y") await game();
```
2번의 실행만 가정하고 동작하고있는것같아요. white, break 를 이용해서 지속적으로 프로그램이 실행되게 만들어볼까요?

onschan commented on src/domain/Lotto.js:
```diff
@@ -0,0 +1,32 @@
+  }
+
+  validate(numbers) {
+    const validator = new LottoValidator();
```
```ts
constructor(numbers, validator = new LottoValidator()) {
    this.validator = validator;
    this.numbers = this.sortLottoNumber(numbers);
    this.validate(this.numbers);
  }
  
  validate(numbers) {
    this.validator.validateLotto(numbers);
  }
```

LottoValidator 클래스가 Lotto 클래스만을 목적으로 제작된 의도를 가진다고 하더라도
클래스 내부에서 직접 인스턴스를 생성하는 케이스는 의존성 주입(DI)를 통해 결합도를 낮춰주는 편이 안정적이고 확장에 유리해요.

onschan reviewed (changes requested):
안녕하세요! 다이앤!

만나서 반가워요 😊

3인 페어를 진행하셨군요 ㅋㅋㅋ! 코드에서도 많은 고민이 오고갔다는게 느껴집니다!

로또라는 도메인을 책임과 역할에 맞게 잘 쪼개려고 노력하시고 잘 반영해주셨어요!

질문 사항에도 있지만 객체의 테스트에 다른 객체가 필요한 케이스가 왜 발생했을지, 이런 의존성과 결합도를 어떻게 낮출 수 있을지에 대해 코멘트를 남겨두었지만 이 부분에 집중해서 고민해보신다면 더 얻어가시는게 있을거라고 생각해요.

---

### 질문에 대한 답변

### 

>온스타는 class/객체/함수를 나누는 기준이 있나요?

class/ literal object / function에 대한 구분 기준이라고 생각할게요!

필드가 필요한지 아닌지로에 대한 기준을 잡아주셨다고 하셨는데 이 부분에 동의합니다. 다만 조금 다르게 표현할게요. 

클래스는 객체를 생성하기위한 설계도와 같습니다. 값, 함수등이 긴밀하게 결합되어 복잡한 동작을 풀어내야할때 혹은 상속을 활용하여 정의된 객체가 확장되어 사용되는 것이 유리할때 클래스를 사용합니다. 

“필드가 필요하다”라는 부분은 전자에 해당되는 것 같아요.

질문의 “객체”라는 부분을 어떻게 이해하시는지 모호하긴한데, Ticket의 구현체를 미루어 보아 literal object를 의미하시는 것 같고 이는 같은 관심사를 묶어두는 정도로 보여요. constants 등의 관리와 동일한 정도의 레벨로 보입니다. 도메인 구현에 사용한다면 Class와 동일하게 구현하고 동작할 수 있지만 굳이 그런 수고를 들이고 싶진 않네요. 😅

function에 대한 질문은 FP에 대한 질문일까요? 이 부분도 살짝 모호하네요. 요구사항이 절차적으로 명확하고 단순하다면 Class 설계 대신 인자와 반환값에 대해 단일 책임만 하는 작은 함수 조각들로 도메인을 쪼개서 표현할 수 있을 거에요. 구조적인 설계가 아닌 단순 구현이 목적인 경우, 해당 방법을 많이 사용합니다.

목적과 구현 방법에 대한 정의와 설계가 필요하다면 클래스로 도메인 설계를 진행하고 구현 위주로 필요한 함수들을 만들어 진행한다면 작은 함수들을 만들어가면서 도메인을 그려낼 것 같아요.

당연히 상황에 따라 적절히 잘 선택하는 편이 좋고, 다이앤이 적어주신 구분해신 기준들은 명확하게 잘 해주셨다고 생각해요 😊

>저희는 역할에 따라 domain을 분리해 봤습니다. 최대한 비슷한 역할을 하는 코드가 같이 있도록 짜봤는데 잘 구분이 되었는지 궁금합니다
`Lotto` : 숫자확인 로직들
`LottoResult` : 당첨내역/수익률 계산
`LottoMachine` : 발행된 로또들 저장, 발행한 로또들을 돌면서 숫자 확인하고 순위 반환

로또라는 도메인을 잘 해석에 그에 따라 잘 나눠주셨다고 생각해요. 다만 Lotto 클래스와 LottoMachine이 강하게 결합되어 있는 부분만 덜어낸다면 좋을 것 같아요. 클래스 내부에서 다른 클래스의 인스턴스를 직접 생성하거나 내부 구현을 알고 사용한다면 이질감을 느끼고 이 의존성을 어떻게 관리해야될지는 고민해보는게 좋아요.

>페어 프로그래밍을 하면서 메소드 선언 순서에 대해 의견이 갈습니다. 저는 코드의 가독성을 높이기 위해, 메소드가 호출되는 흐름에 따라 선언하는 것이 좋다고 생각했습니다. 예를 들어, a 함수와 b 함수가 존재하고, 이 둘을 사용하는 c 함수가 있다면, c 함수를 후순위에 두는 것이 더 자연스럽다고 판단했습니다 다른 페어는 핵심 로직을 위에 작성하는 것이 좋다고 판단했고 논의한 끝에 호출되는 흐름에 따라 쓰기로 결정했습니다. 온스타는 어떻게 생각하시나요?

이 부분도 역시 정답이 없는 부분이라 더 치열하게 논의하셨겠네요 🥹 

아마 step1-index.js를 작성하시면서 논의하신 것 같아요. c를 선순위로 두시겠다고 하셨던건 호이스팅도 생각하셔서 하셨던거겠죠?

우선 코멘트엔 남기지 않았지만 step1-index.js의 함수 선언에 대해서는 모듈 시스템을 먼저 적극적으로 활용하는게 좋을 것 같아요.

어플리케이션의 시작점인 step-index.js에서 필요로 하는 함수를 전부 선언하기보다는 도메인과 UI가 결합되는 컨트롤러 영역으로 지정된 모듈에서 관리가 되는 편이 좋을 것 같아요.

본론으로 돌아가자면 관심사가 모듈 시스템하에 잘 관리되고 있다는 상황이라면 저는 핵심 함수를 코드 상단에 둘 것 같아요. 스코프 오염에 대한 걱정도 없어 호이스팅도 잘 활용할 수 있고 메서드가 작은 역할만 하도록 잘 쪼개져있다면 핵심 함수에서 사용되는 다른 함수들은 구현체를 볼 필요없이 함수명과 맥락만 보고도 이해가 가능해야하거든요. a, b, c 3개의 케이스가 아니라 사용되는 함수가 10개라면 스크롤을 쭉 내려야 그제서야 주인공이 누구인지 확인할 수 있잖아요? 이러면 주인공이 서운해할 것 같네요. (사실 취향의 영역이라 생각하긴합니다.)

>TDD를 통해 작은 단위로 문제를 바라볼 수 잇다는 것은 좋은 점이라고 느꼈습니다. 하지만 도메인에 대해 숙지 되지 않은 상태에서 TDD를 사용하다 보니 시간이 더 오래 걸리고 시간대비 어느 장점이 있는지 와닿지 않는 것 같습니다. 기능 개발 후 테스트를 작성하는 것과의 뚜렷한 장점도 잘 느끼지 못한 것 같습니다.

다른 크루 리뷰에서도 비슷한 말을 했지만, 저도 현업에서 TDD를 적극적으로 사용하고 있지는 않아요. 다만 FE TDD를 도입한 회사들의 사례를 보면, 자동화와 셋팅이 잘 갖춰져 있다면 충분히 빠른 개발 속도와 안정성을 동시에 확보할 수 있다고 하더라고요. 기능 요구 사항 위주로 동작을 위한 테스트 간단한 테스트 코드가 빠르게 생성된다면 효용을 충분히 느낄 수 있을 거라는 것에는 저는 크게 동감해요. 

빠르게 생성되는 테스트 코드가 UX, UI가 긴밀하게 연결되어있는 영역까지 커버리지를 넓히기는 어려운 영역일 수 있겠지만 FE 개발자들의 하나의 숙제라고도 생각해요. 어렵고 효용을 지금은 못느끼실 수 있지만 할 수 있는데 도입을 하지 않는 것과 못하는 것은 차이가 있다고 생각해요. 

우테코처럼 열의가 불타오르고 이것저것 시도해볼 수 있는 시기에 이것저것 시도해본다면 더 확실한 다이앤의 주관이 생길 거고, 혹시라도 생태계에 기여할 수도 있지않을까요?

>첫번째 미션에서도 질문했던 거긴 하지만 아직 단위 테스트의 기준이 뭔지 잘 모르겠습니다. 단위테스트는 독립적으로 테스트가 가능해야 한다고 생각하는데, 지금은 한 객체를 테스트하는데 다른 객체가 필요해서 테스트가 독립적으로 작동하지 않는다고 생각해요. 온스타는 어떻게 생각하시나요?

모듈 단위 테스트를 진행 중인데 다른 모듈에 대해 의존이 생긴다면 그 의존이 어디서 발생하고 있는지 파악해보는 것도 좋을 거에요. 보통 단위테스트에서 떼어낼 수 없는 의존은 모킹을 도입하거나 할 수 있는데 객체간의 의존은 더욱 관리하기 힘들어져요. LottoMachine과 Lotto의 의존의 경우는 코멘트를 통해 제 의견을 남겨두었으니 확인해주세요.

Daeun-100 committed:
[REFACTOR] : 재시작 입력받는 함수 분리 및 게임 반복이 가능하도록 로직 수정

Daeun-100 committed:
[REFACTOR] : validator 테스트에서 [ERROR] 메세지를 상수화된 메세지로 변경

Daeun-100 committed:
[TEST] :  로또 객체의 책임과 무관한 테스트를 분리하여 관련 파일로 이동

Daeun-100 committed:
[TEST] : 로또 등수 판별 테스트를 RANK 값만 검증하도록 변경

Daeun-100 committed:
[REFACTOR] : LottoResult test에 describe 추가

Daeun-100 committed:
[TEST] : 당첨 내역이 없을 경우의 총수입을 구하는 테스트 추가

Daeun-100 committed:
[REFACTOR] : ERROR 메세지의 구입금액 단위와 로또번호 개수를 상수로 변경

Daeun-100 committed:
[REFACTOR] : 구입금액 범위 테스트의 값을 경계값으로 변경

Daeun-100 committed:
[REFACTOR] : 1000단위 구분자 추가

Daeun-100 committed:
[TEST] : 입력값 범위 테스트를 경계값을 기준으로 테스트 하도록 변경

Daeun-100 committed:
[REFACOR] : Lotto 클래스의 number 필드를 private으로 변경

Daeun-100 committed:
[REFACTOR] : LottoMachine 클래스의 issuedLottoNumbers 필드를 private으로 변경

Daeun-100 committed:
[REFACTOR] : checkANdAddLottoNumbers에서 store.has 문 삭제

Daeun-100 committed:
[REFACTOR] : LottoMachine 클래스의 matchedLottoStatus 필드를 private으로 변환

Daeun-100 committed:
[REFACTOR] : LottoMachine 클래스 이름 변경 및 로직 변경

- LottoMachien -> LottoRank 로 이름 변경
- 파라미터는 로또 객체가 아닌 숫자 배열을 받도록 변경
- Lotto 클래스에 있던 숫자 확인 로직을 LottoRank 클래스로 옮김

Daeun-100 committed:
[REFACTOR] : LottoRank 클래스 생성시 계산에 필요한 로또 번호를 파라미터로 받도록 변경

Daeun-100 committed:
[REFACTOR] : LOTTORANK 클래스 이름 변경 및 메소드명 수정

Daeun-100 committed:
[REFACTOR] : 로또 status 계산하는 로직 수정

Daeun-100 committed:
[REFACOT] : LottoRank 테스트에서 Rank 값만 비교하도록 수정

Daeun-100 committed:
[REFACTOR] : LottoResult에서 결과 업데이트 하는 함수 분리 및 constructor에서 바로 실행하도록 수정

Daeun-100 committed:
[REFACTOR] : LottoStatus 클래스 생성시 matchLottoStatus()가 바로 실행되도록 수정

Daeun-100 committed:
[REFACTOR] : 수익률 자릿수 가공 로직을 Output에서 하도록 변경

Daeun-100 committed:
[REFACTOR] : step1-index 로직을 controller 폴더 파일로 분리

Daeun-100 committed:
[REFACTOR] : validate 클래스들을  함수로 변경

Daeun-100 committed:
[REFACTOR] : WINNING_HISTORY 초기값 상수화화

Daeun-100 committed:
[REFACTOR] : vaildateLotto test의 숫자 배열 생성을 LOTTO_NUMBER.LENGTH 상수를 이용하도록 수정

Daeun-100 committed:
[FIX] : REWORD -> REWARD 오타 수정

Daeun-100 committed:
[REFACTOR] : 각 clas에서 필드와 메소드들을 private 로 변경

Daeun-100 committed:
[REFACTOR] : IS_BONUS 가 필요없는 status는 property값을 null로 설정 및 status 반환 로직 수정

Daeun-100 committed:
[CHORE] : LottoRank.test -> LottoStatus.test 이름 수정

Daeun-100 commented:
안녕하세요 온스타! 
온스타의 답변과 리뷰를 보고 유지보수하기 좋은 코드에 대해 살짝 감을 잡은 것 같아요. 해당 로직에 대한 꼼꼼한 테스트 케이스 작성과 분류에 대해서도 다시 한번 생각해 보는 시간이였습니다! 그렇지만 아직도 궁금하고 모르는게 너무 많습니다.🥹 

아래는 리뷰를 바탕으로 리팩토링 한 내용과 리팩토링 과정에서 궁금했던 점들입니다.

---
### ✏️ LottoMachine 클래스 이름 변경 및 Lotto 클래스와의 결합 제거
- LottoMachine 이름은 로또의 status를 반환하는 역할 파악이 힘들다고 생각해 LottoStatus로 이름을 바꿨습니다. https://github.com/woowacourse/javascript-lotto/pull/360/commits/e5ee8ba6ce7efedd0e974e7675f8500d536b4a88
- 또한 Lotto 클래스와의 강한 결합을 없애기 위해 로또의 번호만 받는 형식으로 수정하였습니다https://github.com/woowacourse/javascript-lotto/pull/360/commits/bb8e9234dec77eed744e61b671291fd35de50c8c
###  ❓LottoStatus 클래스의 생성자 파라미터,필드 추가

- https://github.com/woowacourse/javascript-lotto/pull/360/commits/a1bf4bf4b9a162757611d43326c538a07249174a
- 로또의 등수를 알기 위해서는 로또 번호, 당첨 번호, 보너스 번호가 필요하고 당첨 번호와 보너스 번호는 한번 설정되면 바뀌면 안되는 정보라고 생각합니다. 기존 로직에서는 당첨 번호와 보너스 번호를 메소드의 파라미터로 넘겨서 계산을 하고 있어 중간에 잘못된 값이 넘어갈 위험성이 있다고 판단했습니다.  따라서 아예 LottoStatus 객체를 생성할때 로또 번호, 당첨 번호, 보너스 번호를 한번에 전부 넘겨주는 방식으로 구현을 했습니다. 그런데 이런 방식으로 구현을 하니 넘겨주는 파라미터와 내부 필드가 많아지더라고요 😅
    ```
  constructor({ issuedLottoNumbers, enteredLottoNumbers, bonusLottoNumber }) {
    this.#issuedLottoNumbers = issuedLottoNumbers;
    this.#enteredLottoNumbers = enteredLottoNumbers;
    this.#bonusLottoNumber = bonusLottoNumber;
    this.#matchedLottoStatus = [];
    this.#matchLottoStatus();
  }
   ```
   프로그래밍 요구 사항에 `클래스를 사용하는 경우, 3개 이상의 인스턴스 변수를 가진 클래스를 쓰지 않는다.` 가 있어서 필드를 줄이거나 역할을 분리 할 수 있는 방법이 있을까 생각해봤지만 방법이 떠오르지 않아 이대로 구현 을 했습니다. `메소드의 파라미터로 값을 넘겨주기`와 `객체 생성자의 파라 미터로 전달해서 내부 필드에 저장하기` 두 방법이 각각 어떤 상황에서 쓰기 적절한지 아직 잘 모르겠습니다.

### ❓함수를 나누는 기준, util 과 메서드
- 앞서 드린 질문과 비슷한 면도 있고 이전에 올린 PR에서 제 질문이 명확하지 않았던 것 같아 다시 정리해서 여쭤봅니다! 
예를 들어 지금 코드에서 LottoStatus 안에 matchingCount 를 계산하는 함수가 있습니다
   ```
  #countMatchingNumbers(numbers) {
    return numbers.filter((number) =>
      this.#enteredLottoNumbers.includes(number)
    ).length;
  }
   ```
   위 메서드를 enterdNumber를 인자로 받는 유틸 함수로 변경해서 사용 할 수 도 있습니다

   ```
   export function countMatchingNumbers(enteredNumbers, 
   drawnNumbers) {
     return drawnNumbers.filter((number) => 
   enteredNumbers.includes(number)).length;
   }
   ```
   두 방법의 동작은 같지만 클래스 내부 필드를 활용하는 메서드/ 인자를 받아 동작하는 유틸 함수의 차이가 있다고 이해하고 있고 어떤 방식을 선택해야 할지 아직 잘 모르겠습니다


### ✏️ vaildator 클래스를 함수로 변경 

- https://github.com/woowacourse/javascript-lotto/pull/360/commits/362d69baab1aef5d479a6ba0bb77271a7d96049e
- [validate 클래스의 의존성 주입](https://github.com/woowacourse/javascript-lotto/pull/360#discussion_r1964210847) 코멘트를 받고 validator를 리팩토링하다 validator를  굳이 class로 만들어야 하나? 라는 의문이 들었습니다. 기존 validator 클래스들은 기본 validator 클래스를 상속받고 있었는데 수업에서 객체의 상속과 조합에 대해 배우면서  validator 클래스가 다른 LottoValidator,BonusNumberValidator... 등의 클래스의 부모 역할을 하는 것이 적절한가?하는 의문이 들었습니다. 각 클래스들은 validator의 기본 검증 함수를 가져와서 사용할 뿐 상위 계층의 클래스라고 하기에는 애매하다고 판단했습니다. 그러면 조합을 사용할 방법이 있을까 찾아보다가 최종적으로는 validator을 class로 만들 이유가 없다고 판단해 함수로 변경하게 되었습니다. 
### ❓인스턴스를 생성하지 않는 경우의 의존성 주입
- 클래스 내부에서 직접 인스턴스를 생성하는 경우 의존성 주입을 통해 결합도를 낮춰주는 편이 좋다는 것은 이해했습니다! 그런데 바뀐 코드처럼 함수를 가져와 쓰는 경우에도 의존성 주입을 해주는 편이 좋을까요?

### ✏️ 로또의 status 계산하는 로직 수정 
- 기존 로직의 가독성이 떨어진다고 생각해 map 과 filter를 사용하는 로직으로 변경하였습니다. https://github.com/woowacourse/javascript-lotto/pull/360/commits/088f34d3f734a6d7cc3288b0684a6d3852c473ce
-  LOTTO_STATUS의 IS_BONUS 값을 확인할 필요가 없는 곳은 null로 변경 하였습니다. https://github.com/woowacourse/javascript-lotto/pull/360/commits/3cd999bc22e8b07bf0a0ae2c8310b6b8071dda67


|리뷰 내용| 관련 코멘트| 반영 커밋 |
|:---|:---:|:---:|
|정확한 에러 메세지 작성|https://github.com/woowacourse/javascript-lotto/pull/360#discussion_r1963648645|https://github.com/woowacourse/javascript-lotto/pull/360/commits/10852af5a8a613cae9338a6e547719080e4522e2|
|테스트가 Lotto 객체 테스트와 어울리지 않음|https://github.com/woowacourse/javascript-lotto/pull/360#discussion_r1963710437|https://github.com/woowacourse/javascript-lotto/pull/360/commits/b4c1a32f6eb5181ef408f861ffcc3e6c953e4e49|
|describe 추가|https://github.com/woowacourse/javascript-lotto/pull/360#discussion_r1963797952 <br>https://github.com/woowacourse/javascript-lotto/pull/360#discussion_r1963872989|https://github.com/woowacourse/javascript-lotto/pull/360/commits/349a9b6b4b45fd66b7a1340c775f8ed79a2476ff|
|등수 반환 테스트가 세부 요구 사항 변경에 취약|https://github.com/woowacourse/javascript-lotto/pull/360#discussion_r1963809534|https://github.com/woowacourse/javascript-lotto/pull/360/commits/69cb5865afada3474533b7cb1673a5b9a5aafe4f
|도메인 반환값 |https://github.com/woowacourse/javascript-lotto/pull/360#discussion_r1963839920|https://github.com/woowacourse/javascript-lotto/pull/360/commits/533338baaa2e7e4a014b85c3815a5dab5675542e
|당첨내역이 없는 경우의 테스트 추가|https://github.com/woowacourse/javascript-lotto/pull/360#discussion_r1963877776|https://github.com/woowacourse/javascript-lotto/pull/360/commits/dc3327fc3bf59c6c6745a23b1ab40f9335ff81b1
|출력문구의 매직넘버 상수화|https://github.com/woowacourse/javascript-lotto/pull/360#discussion_r1963885269|https://github.com/woowacourse/javascript-lotto/pull/360/commits/1791c330484646daf5137bf0f934a5ae76dd832b
|경계값 기준으로 테스트|https://github.com/woowacourse/javascript-lotto/pull/360#discussion_r1963908825 <br>https://github.com/woowacourse/javascript-lotto/pull/360#discussion_r1963916015| https://github.com/woowacourse/javascript-lotto/pull/360/commits/5ef1c40c7027ae1e929824cc2f1df2823ed5f49a <br>https://github.com/woowacourse/javascript-lotto/pull/360/commits/4037a20afbbbc3b11cc099f6ef042383fb414cb0
|1000 단위 구분자 추가|https://github.com/woowacourse/javascript-lotto/pull/360#discussion_r1963910856| https://github.com/woowacourse/javascript-lotto/pull/360/commits/650b0c5aff55b044e1293e0e98e875d8b9d4df23
| private 필드로 변환|https://github.com/woowacourse/javascript-lotto/pull/360#discussion_r1963959585|https://github.com/woowacourse/javascript-lotto/pull/360/commits/e1975db704fed759198916ade97cac6d9008fbdc<br>https://github.com/woowacourse/javascript-lotto/pull/360/commits/920aaad3f111c33aad952ae8490c8487a2a33bcc<br>https://github.com/woowacourse/javascript-lotto/pull/360/commits/87240de21d35ee06734b8b1f0b5abbab5e4bd779 <br>https://github.com/woowacourse/javascript-lotto/pull/360/commits/4afbe66a92eca088adbab4335b82c70e041570a4
|getter 역할 외 로직 외부 분리|https://github.com/woowacourse/javascript-lotto/pull/360#discussion_r1964017099|https://github.com/woowacourse/javascript-lotto/pull/360/commits/088f34d3f734a6d7cc3288b0684a6d3852c473ce
|set 사용시  중복확인 필요x|https://github.com/woowacourse/javascript-lotto/pull/360#discussion_r1964033673|https://github.com/woowacourse/javascript-lotto/pull/360/commits/f2ce1eb1e437883cebe51939fcce3712c08e6ebb
|retry 함수 모듈화|https://github.com/woowacourse/javascript-lotto/pull/360#discussion_r1964049685|https://github.com/woowacourse/javascript-lotto/pull/360/commits/fed2868c37f9b408ab48255d28e0288ff7193902
|지속 실행 가능하도록 수정|https://github.com/woowacourse/javascript-lotto/pull/360#discussion_r1964050620|https://github.com/woowacourse/javascript-lotto/pull/360/commits/fed2868c37f9b408ab48255d28e0288ff7193902
|index 로직을 controller 폴더로 분리||https://github.com/woowacourse/javascript-lotto/pull/360/commits/262ffd04fc323f479fc0d66499fd6c506be146ff
|WINNING_HISTORY 초기값 상수화||https://github.com/woowacourse/javascript-lotto/pull/360/commits/c2302f49e01edae68b604015d0e4594e9ac35d2e

onschan reviewed (changes requested):
안녕하세요. 다이앤.

성실하게 리뷰 반영해주셔서 감사합니다 ☺️

반영 사항도 표로 잘 정리해서 적어주셨군요!

너무 정성들여주셨는데 코멘트에 직접 반영된 커밋을 적어주시는게 좀 더 직관적일 것 같아요 😅

>❓ LottoStatus 클래스의 생성자 파라미터,필드 추가
그런데 이런 방식으로 구현을 하니 넘겨주는 파라미터와 내부 필드가 많아지더라고요 😅프로그래밍 요구 사항에 `클래스를 사용하는 경우, 3개 이상의 인스턴스 변수를 가진 클래스를 쓰지 않는다.` 가 있어서 필드를 줄이거나 역할을 분리 할 수 있는 방법이 있을까 생각해봤지만 방법이 떠오르지 않아 이대로 구현 을 했습니다. `메소드의 파라미터로 값을 넘겨주기`와 `객체 생성자의 파라 미터로 전달해서 내부 필드에 저장하기` 두 방법이 각각 어떤 상황에서 쓰기 적절한지 아직 잘 모르겠습니다.

좋은 고민이네요! 메서드의 인자로 값을 주입하면 위험성이 있을 수도 있다는 부분도 공감할 수 있는 부분이에요. 지금 구조도 충분히 괜찮아보이지만 `클래스를 사용하는 경우, 3개 이상의 인스턴스 변수를 가진 클래스를 쓰지 않는다.` 라는 제시된 요구 사항을 충족하는 것도 중요한 일이죠.

LottoStatus 클래스는 입력된 로또 번호를 꼭 필드로 가지고있어야될까요?

당첨 번호와 보너스 번호만 알고있다면, 여러개의 입력된 로또 번호의 결과를 반환하는 시스템처럼 사용할 수 있을 것 같은데요.

```tsx
const matchedLottoStatus1 = lottoStatus.getMatchedLottoStatus([[[1, 2, 3, 4, 5, 6], [2, 3, 4, 5, 6, 7]]);
const matchedLottoStatus2 = lottoStatus.getMatchedLottoStatus([[2, 3, 4, 5, 6, 7]]);
const matchedLottoStatus3 = lottoStatus.getMatchedLottoStatus([[3, 4, 5, 6, 7, 8]]);
```

이런 방향은 어떻게 생각하시나요?

>❓ 함수를 나누는 기준, util 과 메서드
두 방법의 동작은 같지만 클래스 내부 필드를 활용하는 메서드/ 인자를 받아 동작하는 유틸 함수의 차이가 있다고 이해하고 있고 어떤 방식을 선택해야 할지 아직 잘 모르겠습니다

countMatchingNumbers 메서드(두 배열 중 교집합을 카운팅 해주는 동작)는 LottoStatus의 고유의 동작일까요?

어떤 해석을 하더라도 정답과 오답은 없습니다만, 클래스는 객체의 설계라는 시각에서 바라본다면 정말 필요한 필드와 메서드만 설계에 포함되는 편이 좋아보여요. 

countMatchingNumbers라는 메서드 없이 getMatchingCounts 메서드에 countMathchingNumber의 코드 조각이 그대로 들어가도 해당 클래스의 동작을 이해하는데는 문제가 없을것 같습니다. 이 코드 조각을 하나의 함수 혹은 메서드로 분리하고 싶다는 니즈로 countMathchingNumber 메서드가 추가된 것 같구요. 이 시각으로 본다면 countMathchingNumber는 유틸 함수쪽에 가까워보여요.

외부 함수로 분리해도될까, 메서드로 포함해야되나라고 고민이 들때는 이 클래스의 고유의 동작일지에 대해 고민해보시면 좋을 것 같아요.

>❓ 인스턴스를 생성하지 않는 경우의 의존성 주입
클래스 내부에서 직접 인스턴스를 생성하는 경우 의존성 주입을 통해 결합도를 낮춰주는 편이 좋다는 것은 이해했습니다! 그런데 바뀐 코드처럼 함수를 가져와 쓰는 경우에도 의존성 주입을 해주는 편이 좋을까요?

인스턴스의 경우, 어떤 메서드를 가지고있는지 그 메서드와 연관된 인스턴스의 내부 필드는 어떻게 관여되는지, 어떤 사이드이펙트가 발생할 수 있는지 예측하기 어려워요. 그렇기때문에 “인스턴스는 조금 더 신경써서 관리해주면 좋다.” 정도의 의견이었어요. 이미 생성과정에서 강결합이 의도되어 생성된 클래스였고 함수로 바뀐 상황에서도 강결합이 예상되는 상황이에요. 굳이 DI를 해줄 필요는 현재 구현 사항안에서 필요는 없다는 의견이에요.

다만, 인스턴스의 주입이든, 함수의 주입이든 가져갈 수 있는 장점은 동일해요. 결합도를 낮추고 니즈에 따라 사용처에서 바꿔가면서 사용할 수 있으니깐요.

확장성에 더 신경쓰고 싶다라면 함수의 주입을 해도 좋습니다만 항상 필요로 하는 건 아니니깐 도메인의 성격이나 도메인이 얼마나 변동성이 높은지도 잘 고려해서 설계시에 결정하면 될 것 같습니다. 결합도가 낮은건 좋은거니깐 무조건 적용해야지! 라는 생각보다는 점점 고민들을 토대로 설계를 진행하시다보면 음~ 이 부분엔 적용하는게 유리하겠네~ 를 판단할 수 있게될 거에요. 저도 그 경지에는 다다르지 못했지만 본인의 기준을 늘 세우는건 중요합니다!

---
수정 사항도 잘 반영해주셨고 이대로 머지해도 되겠지만,
`클래스를 사용하는 경우, 3개 이상의 인스턴스 변수를 가진 클래스를 쓰지 않는다.`
해당 요구 사항은 어떻게 충족하면 좋을지 조금 더 고민해보시고 반영해보시면 좋을 것 같아요.
문제가 있어서 말씀드리는건 아니니 짧게 고민해주시고 반영해보신 후 리뷰 요청해주세요!

Daeun-100 committed:
[REFACTOR] : countMatchingNumbers를 유틸 함수로 분리

Daeun-100 committed:
[REFACTOR] : LottoStatus에서 issuedLottoNumbers 파라미터와 필드 삭제, matchLottoStatus 필드 삭제

Daeun-100 committed:
[REFACTOR] : 불필요한 중괄호와 return 제거

Daeun-100 commented:
상세한 답변 감사합니다 온스타!😊

**https://github.com/woowacourse/javascript-lotto/pull/360/commits/a4d54c1554c1dee198773631e807a59522fb4f73 issuedLottoNumbers,matchedLottoStatus 필드 삭제**

LottoStatus 객체에서 중요한건 당첨번호와 보너스번호라 생각해 발행된 로또 번호는 생성자 파라미터와 필드에서 삭제하고 메서드의 파라미터로 받게 수정하였습니다. issuedLottoNumbers 필드가 삭제됨에 따라 matchedLottoStatus 필드도 삭제하고 바로 계산된 status들을 반환하게 수정했습니다

**https://github.com/woowacourse/javascript-lotto/pull/360/commits/5e9fa0b042ed4398d5d63820c8e44b1bcb913c48 countMatchingNumbers를 유틸함수로 분리하였습니다**
**https://github.com/woowacourse/javascript-lotto/pull/360/commits/0b933ddee513c47142a83b954db12962618cb49c 함수에서 불필요한 { return } 문을 생략하였습니다**

onschan commented on src/utils/countMatchingNumbers.js:
```diff
@@ -0,0 +1,16 @@
+const countMatchingNumbers = (numberArr1, numberArr2) => {
+  const arr1 = [...numberArr1];
+  const arr2 = [...numberArr2];
```
요건 왜 필요한건가요?

onschan commented on src/utils/countMatchingNumbers.js:
```diff
@@ -0,0 +1,16 @@
+  const arr1 = [...numberArr1];
+  const arr2 = [...numberArr2];
+
+  //배열안의 내용이 숫자가 아닌경우 예외처리
+  if (
+    !arr1.every((element) => typeof element === "number") ||
+    !arr2.every((element) => typeof element === "number")
+  ) {
+    throw new Error("숫자 배열이 아닙니다.");
+  }
```
utils 함수의 이런 디테일 챙기는 부분 너무 좋습니다 👍👍

`if ([...numberArra1, ...numberArr2].some(typeof element !== "number"])`

some 을 이용하는게 성능적으로도 직관적으로도 이해하기 쉬울 것 같아요!

onschan reviewed (approved):
바로 반영해주셨네요 다이앤!

이번 미션은 여기서 완료하고 다음 단계에서 뵐까합니다!
성실히 잘 반영해주셔서 감사해요 ☺️
